go/processor/floria/processor.go

```diff
@@ -70,31 +76,74 @@ func (p *processor) Run(
 		// Create new contract
 	} else {
 		// Call existing contract
-		result, err = call(p.interpreter, transaction, context)
+		result, err = call(p.interpreter, transaction, context, gas)
 		if err != nil {
-			return errorReceipt, nil
+			return errorReceipt, err
 		}
 	}
 
+	gasUsed := gasUsed(transaction, result.GasLeft)
+
 	return tosca.Receipt{
 		Success:         result.Success,
-		GasUsed:         transaction.GasLimit,
+		GasUsed:         gasUsed,
 		ContractAddress: nil,
 		Output:          result.Output,
 		Logs:            nil,
 	}, nil
 }
 
+func gasUsed(transaction tosca.Transaction, gasLeft tosca.Gas) tosca.Gas {
+	// 10% of remaining gas is charged for non-internal transactions
+	if transaction.Sender != (tosca.Address{}) {
+		gasLeft -= gasLeft / 10
+	}
+
+	return transaction.GasLimit - tosca.Gas(gasLeft)
```

Comment:
nit: it should be fine like this, since `gasLeft` is already of type `tosca.Gas`
```suggestion
	return transaction.GasLimit - gasLeft
```